File: packages/examples/todo/backends/rust/src/commands/update.rs
use super::{not_found, ok, schema, Command};
use crate::store::{Changes, TodoStore};
use crate::types::Priority;
use afd::{failure, CommandError, CommandResult};
use serde::Deserialize;
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Update(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {
    id: String,
    title: Option<String>,
    description: Option<String>,
    priority: Option<Priority>,
    completed: Option<bool>,
}

impl Command for Update {
    type Input = Input;
    const NAME: &'static str = "todo-update";
    const DESCRIPTION: &'static str = "Update a todo's title, description, priority or status";
    const MUTATION: bool = true;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "id": schema::id(),
                "title": schema::title(),
                "description": schema::description(),
                "priority": schema::priority("New priority"),
                "completed": { "type": "boolean", "description": "New completion status" }
            },
            "required": ["id"]
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        let changes = Changes {
            title: input.title,
            description: input.description,
            priority: input.priority,
            completed: input.completed,
        };
        if changes.is_empty() {
            return failure(
                CommandError::new("NO_CHANGES", "No fields to update")
                    .with_suggestion(
                        "Provide at least one of: title, description, completed, priority",
                    )
                    .with_retryable(false),
            );
        }

        match self.0.update(&input.id, changes) {
            Some(todo) => {
                let reasoning = format!("Updated todo \"{}\"", todo.title);
                ok(&todo, reasoning, 1.0, Vec::new())
            }
            None => failure(not_found(&input.id)),
        }
    }
}

File: packages/examples/todo-directclient/backend/src/commands/update.ts
/**
 * @fileoverview todo-update command
 */

import { defineCommand, failure, success } from '@lushly-dev/afd-server';
import { z } from 'zod';
import { store } from '../store/index.js';
import type { Todo } from '../types.js';

const inputSchema = z.object({
	id: z.string().min(1, 'Todo ID is required'),
	title: z.string().min(1).max(200).optional(),
	description: z.string().max(1000).optional(),
	completed: z.boolean().optional(),
	priority: z.enum(['low', 'medium', 'high']).optional(),
});

export const updateTodo = defineCommand<typeof inputSchema, Todo>({
	name: 'todo-update',
	description: 'Update a todo item',
	category: 'todo',
	tags: ['todo', 'update', 'write', 'single'],
	mutation: true,
	version: '1.0.0',
	input: inputSchema,
	errors: ['NOT_FOUND', 'NO_CHANGES'],

	async handler(input) {
		const { id, ...updates } = input;

		// Check if there are any changes
		const hasChanges = Object.values(updates).some((v) => v !== undefined);
		if (!hasChanges) {
			return failure({
				code: 'NO_CHANGES',
				message: 'No fields to update',
				suggestion: 'Provide at least one of: title, description, completed, priority',
			});
		}

		const existing = store.get(id);
		if (!existing) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${id}" not found`,
				suggestion: 'Use todo-list to see available todos',
			});
		}

		const updated = store.update(id, {
			title: updates.title,
			description: updates.description,
			completed: updates.completed,
			priority: updates.priority,
		});

		if (!updated) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${id}" not found`,
				suggestion: 'Use todo-list to see available todos',
			});
		}

		// Build change summary
		const changes: string[] = [];
		if (updates.title) changes.push(`title to "${updates.title}"`);
		if (updates.description !== undefined) changes.push('description');
		if (updates.priority) changes.push(`priority to ${updates.priority}`);

		return success(updated, {
			reasoning: `Updated ${changes.join(', ')} for todo "${updated.title}"`,
			confidence: 1.0,
		});
	},
});
